### PR TITLE
Fix the issue in matmul_add_fusion

### PR DESCRIPTION
**Pull request by HectorSVC**
If Matmul + Add has shape [K] * [K, N], reset it to [1, K] * [K, N] will make the output shape to [1, N] will also requires a reshape on the output.
Fix: just remove the shape reset to not fuse it.

### CODE REVIEW

**Comment by askhade** [comment]
Did you consider adding the reshape for output instead? Does the cost incurred from reshaping the input and output out weigh the gains from this fusion? If yes then simply removing the support for this case makes sense... If no then it may be worth it to just add the reshape

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 1]
```diff
@@ -541,6 +541,23 @@ TEST(GraphTransformationTests, MatMulAddFusion_three_input) {
   ASSERT_TRUE(op_to_count["Gemm"] == 1);
 }
 
+TEST(GraphTransformationTests, MatMulAddFusion_negitive_case) {
```
Comment (askhade) [inline onnxruntime/test/optimizer/graph_transform_test.cc]: >MatMulAddFusion_negitive_case [](start = 31, length = 29)

Can you make the name more explicit or add a comment stating what this is exactly testing, more specifically which negative case this test is testing.

**Comment by HectorSVC** [comment]
> Did you consider adding the reshape for output instead? Does the cost incurred from reshaping the input and output out weigh the gains from this fusion? If yes then simply removing the support for this case makes sense... If no then it may be worth it to just add the reshape

There's only very limit gain to to fuse Matmul+Add to Gemm because the basic implementation is almost same. It saved one turn round of kernel call. Add another Reshape will bring in the extra kernel call back which is not worth it.

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 2]
```diff
@@ -541,6 +541,27 @@ TEST(GraphTransformationTests, MatMulAddFusion_three_input) {
...
 }
 
+// Matmul+Add with shape [k]*[k,N]+[N], won't do the fusion
+// We can do the fusion by changing shape to [1,k]*[k,N]+[1,N], then add a reshape [1,N]=>[N]
+// This will bring extra cost. And there's only very limited gain to fuse Matmul+Add to Gemm
+// Since the basic implementation is almost same
+TEST(GraphTransformationTests, MatMulAddFusion_negitive_case) {
+  auto model_uri = MODEL_FOLDER "matmul_add_fusion/3Input/neg_model.onnx";
+
+  std::shared_ptr<Model> p_model;
```
Comment (yuslepukhin) [inline onnxruntime/test/optimizer/graph_transform_test.cc]: >shared_ptr [](start = 7, length = 10)

Do we actually load the model into shared_ptr? Has Yuan seen this?

Comment (HectorSVC) [reply onnxruntime/test/optimizer/graph_transform_test.cc]: we do have the interface use the shared_ptr I'm not sure the reason. Let's have another PR to clean up it if you have any concern.
[https://github.com/microsoft/onnxruntime/blob/aa7c79eac9875ba3b5c9444b13d1878560415ea7/onnxruntime/core/graph/model.cc#L378](https://github.com/microsoft/onnxruntime/blob/aa7c79eac9875ba3b5c9444b13d1878560415ea7/onnxruntime/core/graph/model.cc#L378)

---
In reply to: [347538251](https://github.com/microsoft/onnxruntime/pull/2407#discussion_r347538251) [](ancestors = 347538251)

**Pull request merged**